server/src/main/java/org/elasticsearch/health/node/DiskHealthIndicatorService.java

```diff
@@ -0,0 +1,469 @@
+                 */
+                Set<String> allImpactedIndices = Stream.concat(indicesWithBlock.stream(), indicesOnImpactedNodes.stream())
+                    .collect(Collectors.toSet());
+                healthIndicatorResult = getResultForBlockedIndicesOrRedDataNodes(
+                    impactedNodes,
+                    allImpactedIndices,
+                    atLeastOneRedDataNode,
+                    details
+                );
+            } else if (HealthStatus.RED.equals(healthStatusFromNodes)) {
+                final Set<String> nodesReportingRed = getNodesReportingStatus(diskHealthInfoMap, HealthStatus.RED);
+                Set<DiscoveryNodeRole> rolesOnRedNodes = getRolesOnNodes(nodesReportingRed, clusterState);
+                // First check the highest priority: we had a data node that reported RED:
+                if (rolesOnRedNodes.stream().anyMatch(DiscoveryNodeRole::canContainData)) {
+                    Set<String> indicesOnRedNodes = getIndicesForNodes(nodesReportingRed, clusterState);
+                    healthIndicatorResult = getResultForBlockedIndicesOrRedDataNodes(
+                        getNodesWithDataRole(nodesReportingRed, clusterState),
+                        indicesOnRedNodes,
+                        true,
+                        details
+                    );
+                } else {
+                    healthIndicatorResult = getResultForNonDataNodeProblem(
+                        rolesOnRedNodes,
+                        nodesReportingRed,
+                        details,
+                        healthStatus,
+                        clusterState
+                    );
+                }
+            } else {
+                final Set<String> nodesReportingYellow = getNodesReportingStatus(diskHealthInfoMap, HealthStatus.YELLOW);
+                Set<DiscoveryNodeRole> rolesOnYellowNodes = getRolesOnNodes(nodesReportingYellow, clusterState);
+                // First check the highest priority: we had a data node that reported YELLOW:
+                if (rolesOnYellowNodes.stream().anyMatch(DiscoveryNodeRole::canContainData)) {
+                    healthIndicatorResult = getResultForYellowDataNodes(
+                        getNodesWithDataRole(nodesReportingYellow, clusterState),
+                        details,
+                        healthStatus,
+                        clusterState
+                    );
+                } else {
+                    healthIndicatorResult = getResultForNonDataNodeProblem(
+                        rolesOnYellowNodes,
+                        nodesReportingYellow,
+                        details,
+                        healthStatus,
+                        clusterState
+                    );
+                }
+            }
+        }
+        return healthIndicatorResult;
+    }
+
+    private Set<String> getNodesReportingStatus(Map<String, DiskHealthInfo> diskHealthInfoMap, HealthStatus status) {
```

Comment:
is this reporting node names or node ids? 
would this method benefit from being static and unit tested?